### PR TITLE
cloud/node-controller use InstanceMetadataByProviderID

### PR DESCRIPTION
**Pull request by gongguan**
/kind feature

Ref [#90652](https://github.com/kubernetes/kubernetes/issues/90652)

```release-note
Cloud node-controller use InstancesV2
```

### CODE REVIEW

**Comment by gongguan** [comment]
I had to break current tests because `updateNodeAddress` will judge instance existence by fakeCloud's  ExistsByProviderID
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L232
while `getNodeModifiersFromCloudProvider` won't judge instance existence
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L438
This difference can't be perceived by `InstanceExistsByProviderID`.

BTW, I have no idea whether the sole [ensureNodeExistsByProviderID](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_lifecycle_controller.go#L151) should be replaced by InstanceMetadataByProviderID to remove all `InstanceExistsByProviderID`.

**Comment by gongguan** [comment]
> I had to break current tests because `updateNodeAddress` will judge instance existence by fakeCloud's ExistsByProviderID
> https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L232
> while `getNodeModifiersFromCloudProvider` ignores instance existence
> https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L438
> This difference can't be perceived by `InstanceExistsByProviderID` except judge instance existence in `updateNodeAddress` at the same time.
> 
> BTW, I'm not sure whether the sole [ensureNodeExistsByProviderID](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_lifecycle_controller.go#L151) should be replaced by InstanceMetadataByProviderID to remove all `InstanceExistsByProviderID`.

ping @wojtek-t for thought

File: pkg/controller/cloud/node_controller.go [hunk 1]
```diff
@@ -229,21 +229,16 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
 		return
 	}
 	// Node that isn't present according to the cloud provider shouldn't have its address updated
-	exists, err := ensureNodeExistsByProviderID(ctx, instances, node)
+	instanceMeta, err := instances.InstanceMetadataByProviderID(ctx, node.Spec.ProviderID)
```
Comment (andrewsykim) [inline pkg/controller/cloud/node_controller.go:232]: The Instances interface is pretty bloated IMO. Wondering if it would be worthwhile to add an `InstancesV2` interface that is just `InstanceMetadataByProviderID` for now. If a cloud provider "implements" the v2 interface, we use the new code path here. Otherwise fallback to the existing one. Thoughts? cc @cheftako

Comment (andrewsykim) [reply pkg/controller/cloud/node_controller.go:232]: Started a discussion on the mailing list for this https://groups.google.com/d/msg/kubernetes-sig-cloud-provider/IYbzlMWvndQ/VgXxgGIFAQAJ

Comment (gongguan) [reply pkg/controller/cloud/node_controller.go:232]: Thanks, I have seen it.

Comment (gongguan) [reply pkg/controller/cloud/node_controller.go:232]: Hi @andrewsykim , how's it going? If we are not ready to deprecate NodeAddressesByProviderID, InstanceTypeByProviderID and InstanceExistsByProviderID, would [#90894](https://github.com/kubernetes/kubernetes/pull/90894) be reverted?

Comment (wojtek-t) [reply pkg/controller/cloud/node_controller.go:232]: I didn't initially realize the full intention here - but basically what you meant is that the existing implementation will cover both the old and the new interface. Which is cool.

And eventually, we will (hopefully) remove the old interface (and together with that the methods that are no longer needed).

>  I don't think we need to revert #90894, we can just do the change here.

@andrewsykim - the question then is that if we don't want to revert the part of that original PR adding the new method to the interface (and leave only the implementation addions).
We probably should, right?

Comment (andrewsykim) [reply pkg/controller/cloud/node_controller.go:232]: > @andrewsykim - the question then is that if we don't want to revert the part of that original PR adding the new method to the interface (and leave only the implementation addions).
We probably should, right?

Yeah sorry I wasn't clear. We should revert the addition of `InstanceMetadataByProviderID` and the deprecation of the other methods in `Instances`. I just meant we didn't have to revert the PR, we can just revert the changes here. Either way works for me :).

**Comment by gongguan** [comment]
We are not ready to replace InstanceTypeByProviderID, InstanceExistsByProviderID and NodeAddressesByProviderID by InstanceMetadataByProviderID, so I close this pr.

**Comment by wojtek-t** [comment]
> We are not ready to replace InstanceTypeByProviderID, InstanceExistsByProviderID and NodeAddressesByProviderID by InstanceMetadataByProviderID, so I close this pr.

Actually - we are ready. The approach is just slightly different.

You need to create a new "InstanceV2" interface (the implementations are already done) with a subset of methods and in the controller have switch: if the provider implements new interface then use new code-path, if it doesn't use the old codepath.

**Comment by gongguan** [comment]
> Actually - we are ready. The approach is just slightly different.
> 
> You need to create a new "InstanceV2" interface (the implementations are already done) with a subset of methods and in the controller have switch: if the provider implements new interface then use new code-path, if it doesn't use the old codepath.

I used to thought InstanceV2 is a long-term work so I had closed it. Is InstanceV2 a different package to current Instance?

**Comment by andrewsykim** [comment]
@gongguan here's an example https://github.com/kubernetes/kubernetes/compare/master...andrewsykim:instances-v2?expand=1

**Comment by gongguan** [comment]
> @gongguan here's an example https://github.com/kubernetes/kubernetes/compare/master...andrewsykim:instances-v2?expand=1

Thanks @andrewsykim , I can do follow-up work based your commit.

**Comment by gongguan** [comment]
Hi @andrewsykim , I have read your example code today. And I have a doubt about InstanceV2, it makes node controllers logic tedious. However cnc.cloud.InstancesV2() must exist, right? Can we introduce InstanceMetadataByProviderID to Instance without changing(deprecating) other methods?

**Comment by wojtek-t** [comment]
>  However cnc.cloud.InstancesV2() always exist, right?

It always exists for built-in providers. But there can be other.

> Can we introduce InstanceMetadataByProviderID to Instance without changing(deprecating) other methods?

You effectively did that already.

**Comment by gongguan** [comment]
> It always exists for built-in providers. But there can be other.

Yes, I haven't realized it.

> You effectively did that already.

Sorry, I'm not fully understand it. I just think using current Instance is cleaner because Instance and InstanceV2 two implementation is tedious. Is there other reasons that we should use these two Instance interface?

**Comment by gongguan** [comment]
> Is InstanceMetadataByProviderID(in current Instance) will affect providers not built-in?

Updated, I mean introduce InstanceMetadataByProviderID to current Instance Interface and use InstanceMetadataByProviderID in cloud node controller.

**Comment by andrewsykim** [comment]
We can't replace half of the methods in Instances in favor of `InstanceMetadataByProviderID`, that would be really disruptive to implementations of the interface that have well tested/vetted implementations based on those methods. The call to `InstanceMetadataByProviderID` should be opt-in, at least for a few releases. Adding a InstancesV2 allows for providers to opt-in or opt-out of this optimization. 

The node controller logic doesn't actually have to be so tedious, I updated my example to be a bit cleaner, PTAL. https://github.com/kubernetes/kubernetes/compare/master...andrewsykim:instances-v2?expand=1

**Comment by gongguan** [comment]
> that would be really disruptive to implementations of the interface that have well tested/vetted implementations based on those methods.

It sounds reasonable. 
New commit looks good, but initializeNode and initializeNodeV2 would have many similar logics. Anyway, I'll try to do it.

**Comment by gongguan** [comment]
Updated, PTAL @wojtek-t @andrewsykim 
Made instanceV2 and its implementations in a separated pr #92367

Commented some test cases because of https://github.com/kubernetes/kubernetes/pull/91319#issuecomment-632069759 .
Can `getNodeModifiersFromCloudProvider` not honor instance existence?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go [hunk 2]
```diff
@@ -708,6 +1226,87 @@ func Test_AddCloudNode(t *testing.T) {
...
+					Key:    cloudproviderapi.TaintExternalCloudProvider,
+					Value:  "true",
+					Effect: v1.TaintEffectNoSchedule,
+				},
+			},
+		},
+	}
+
+	fakeCloud := &fakecloud.Cloud{
+		EnableInstancesV2: false,
```
Comment (gongguan) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:1258]: It would be `EnableInstancesV2: true`, I will address it later(together with other feedbacks).

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 3]
```diff
@@ -146,10 +146,13 @@ func (cnc *CloudNodeController) Run(stopCh <-chan struct{}) {
...
 // UpdateNodeStatus updates the node status, such as node addresses
 func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) {
-	instances, ok := cnc.cloud.Instances()
+	var instances cloudprovider.Instances
+	instancesV2, ok := cnc.cloud.InstancesV2()
 	if !ok {
-		utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
-		return
+		if instances, ok = cnc.cloud.Instances(); !ok {
+			utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
```
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:153]: If we are requiring cloud providers to implement the InstancesV2 method, then this change clearly requires a release note.

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:153]: We're not requiring it, but it should be in the release notes anyways.

Comment (cheftako) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:153]: I believe having node controller fail if instanceV2 is not enabled(ok), is effectively requiring instanceV2.

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 4]
```diff
@@ -49,6 +49,8 @@ type Interface interface {
 	LoadBalancer() (LoadBalancer, bool)
 	// Instances returns an instances interface. Also returns true if the interface is supported, false otherwise.
```
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/cloud.go:50]: Can we add a deprecation note for instances? Would like to minimize how long we need to support both.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go [hunk 5]
```diff
@@ -659,6 +659,11 @@ func (g *Cloud) Instances() (cloudprovider.Instances, bool) {
 	return g, true
 }
 
+// InstancesV2 returns an implementation of InstancesV2 for Google Compute Engine.
+func (g *Cloud) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	return g, true
```
Comment (cheftako) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: I don't believe the gcp cloud provider should be returning true as it does not yet implement InstanceShutdownByProviderID. (https://github.com/kubernetes/kubernetes/blob/440b5c2fbc82f4bba3f5d3653ffbd77dec481aa0/staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go#L174) Have we verified that the rest of the in-tree providers actually fully implement the interface?

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 4 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go:50]: @andrewsykim proposed only cloud node-controller use instancesV2, I don't think he want deprecate current Instances.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go (hunk 5 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: > I don't believe the gcp cloud provider should be returning true as it does not yet implement InstanceShutdownByProviderID.

It's consistent with Instances interface.

> Have we verified that the rest of the in-tree providers actually fully implement the interface?

Yes, others in-tree providers all implement InstanceShutdownByProviderID.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 3 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:153]: Updated release note.

**Comment by gongguan** [comment]
@andrewsykim @cheftako Updated, PTAL.
I'm also confused why `getNodeModifiersFromCloudProvider` doesn't check node exists like `updateNodeAddress`.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 6]
```diff
@@ -462,6 +484,12 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(ctx context.Co
 		})
 	}
 
+	// If user provided an IP address, ensure that IP address is found
+	// in the cloud provider before removing the taint on the node
+	if nodeIP, ok := ensureNodeProvidedIPExists(node, nodeAddresses); ok && nodeIP == nil {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:489]: Why was this added here?

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:489]: Same logic of current line 445, because both instances and instancesV2 should handle nodeAddresses by ensureNodeProvidedIPExists.

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:489]: Do you mean it should be checked before `if instanceType != ""` ?

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:489]: I see, I misread the diff sorry :)

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go [hunk 7]
```diff
@@ -675,6 +682,517 @@ func Test_AddCloudNode(t *testing.T) {
...
+			},
+			updatedNode: &v1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "node0",
+					CreationTimestamp: metav1.Date(2012, 1, 1, 0, 0, 0, 0, time.UTC),
+				},
+			},
+		},
+		//{
+		//	name: "[instanceV2] zone/region topology labels added",
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:778]: Are these TODO tests since the fake cloud provider doesn't implement v2?

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 8]
```diff
@@ -221,27 +224,38 @@ func (cnc *CloudNodeController) reconcileNodeLabels(nodeName string) error {
...
-		klog.V(4).Infof("The node %s is no longer present according to the cloud provider, do not process.", node.Name)
-		return
-	}
 
-	nodeAddresses, err := getNodeAddressesByProviderIDOrName(ctx, instances, node.Spec.ProviderID, node.Name)
-	if err != nil {
-		klog.Errorf("Error getting node addresses for node %q: %v", node.Name, err)
-		return
+	var nodeAddresses []v1.NodeAddress
+	if instancesV2 != nil {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:236]: I think we can make this and the code in getNodeModifiersFromCloudProvider a bit cleaner by having one method that returns InstanceMetadata. If `instancesV2` is implemented, construct `InstanceMetadata` from `InstanceMetadataByProviderID`, if `instances` is implemented, construct `InstanceMetadata` by calling the individual methods in `Instances`.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go (hunk 7 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:778]: Reason is https://github.com/kubernetes/kubernetes/pull/91319#issuecomment-647309227

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:778]: Not sure whether to change expected node to make it different from current tests which use instances.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 8 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:236]: Adopted your suggestion and implemented a method getInstanceMetadata which used by getNodeModifiersFromCloudProvider. While getInstanceTypeByProviderIDOrName isn't called here. A redundant call will be produced if replace these code with getInstanceMetadata.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 9]
```diff
@@ -146,10 +146,13 @@ func (cnc *CloudNodeController) Run(stopCh <-chan struct{}) {
 
 // UpdateNodeStatus updates the node status, such as node addresses
 func (cnc *CloudNodeController) UpdateNodeStatus(ctx context.Context) {
-	instances, ok := cnc.cloud.Instances()
+	var instances cloudprovider.Instances
+	instancesV2, ok := cnc.cloud.InstancesV2()
 	if !ok {
-		utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
-		return
+		if instances, ok = cnc.cloud.Instances(); !ok {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:152]: Don't love the nested ifs here, can we do something like this instead:
```go
instancesV2, instanceV2Supported := cnc.cloud.InstancesV2()
instances, instancesSupported := cnc.cloud.Instances()

if instanceV2Supported == nil && instancesSupported == nil {
    utilruntime.HandleError("either Instances or InstanceV2 must be supported")
    return
}
```

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 8 above)
Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:236]: Can we have a separate method for the addresses then? Want to hide the instance / instance v2 checks from the main code path and keep that hidden in separate methods.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 9 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:152]: Sure, looks good.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 8 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:236]: I also want to hide instances/instancesV2 checks in `initializeNode`, however updateNodeAddress and getNodeModifiersFromCloudProvider's logic are slightly different. Do you mean `updateNodeAddress` and `getNodeModifiersFromCloudProvider` using different method to get instanceMetadata(in instances() case)?

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:236]: >  Do you mean updateNodeAddress and getNodeModifiersFromCloudProvider using different method to get instanceMetadata(in instances() case)?

yes, at least for now.

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go]: Updated, PTAL.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 10]
```diff
@@ -221,29 +215,51 @@ func (cnc *CloudNodeController) reconcileNodeLabels(nodeName string) error {
...
-	// Node that isn't present according to the cloud provider shouldn't have its address updated
-	exists, err := ensureNodeExistsByProviderID(ctx, instances, node)
-	if err != nil {
-		// Continue to update node address when not sure the node is not exists
-		klog.Errorf("%v", err)
-	} else if !exists {
-		klog.V(4).Infof("The node %s is no longer present according to the cloud provider, do not process.", node.Name)
-		return
+
+	instanceMetadataGetter := func(providerID string, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:226]: nice :)

**Review by andrewsykim — COMMENTED** [review]
This is looking much better. @gongguan do you think we can get the instanceV2 based unit tests working, would feel more confident about this with some unit tests.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go [hunk 11]
```diff
@@ -390,7 +397,7 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
 
 		// After adding, call UpdateNodeAddress to set the CloudProvider provided IPAddresses
 		// So that users do not see any significant delay in IP addresses being filled into the node
-		cnc.updateNodeAddress(ctx, curNode, instances)
+		cnc.updateNodeAddress(ctx, curNode)
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:400]: Can we actually go one step further. With the current change, you still be calling GetInstanceMetadata twice:
- in getNodeModifiersFromCloudProvider
- here

Can we actually call GetInstanceMetadata in this function and pass it to those two functions?

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:400]: Yes, I have thought it. But I plan to do it as a follow-up work, since this is large enough and tests need to be addressed.

Comment (wojtek-t) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:400]: OK, cool. Can you please add a TODO so that we won't forget about it?

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:400]: Sure.

**Comment by gongguan** [comment]
> This is looking much better. @gongguan do you think we can get the instanceV2 based unit tests working, would feel more confident about this with some unit tests.

@andrewsykim these added test cases for instancesV2 are same as current instances, but I have commented some tests because of failed. This is a more tricky thing I thought. As I said in former comment. Current updateNodeAddress checks instance exists while getNodeModifiersFromCloudProvider doesn't check instance exists. But instanceV2 we have check or not check instance exist in InstanceMetadataByProviderID. That will makes result different. Do you have some suggestions?

**Comment by gongguan** [comment]
Two solutions from me:

- check instance existence in getNodeModifiersFromCloudProvider instances case.
- change instancesV2 's updatedNode in test cases.

Why one do you think better?

**Comment by andrewsykim** [comment]
@gongguan let's remove the exists check in `updateNodeAddress`, please keep it in a separate commit so it's easier to revert.

**Review by andrewsykim — COMMENTED** [review]
@gongguan let's remove the exists check in updateNodeAddress, please keep it in a separate commit so it's easier to revert.

**Comment by feiskyer** [comment]
@andrewsykim @gongguan Do we still plan to get this merged in v1.19?

**Comment by gongguan** [comment]
> @andrewsykim @gongguan Do we still plan to get this merged in v1.19?

I'll update this pr soon, but it's for @andrewsykim to decide.

**Comment by andrewsykim** [comment]
Yes I would like to get this into v1.19, pending unit tests

**Comment by gongguan** [comment]
> @gongguan let's remove the exists check in updateNodeAddress, please keep it in a separate commit so it's easier to revert.

@andrewsykim Updated, removing this check in last pr(not sure whether it should a pre pr). Let wait for pending CI.

File: staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go (hunk 11 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:393]: Added.

**Comment by gongguan** [comment]
I'd like to remove these broken test, because most of them are instance not exist, WDYT @andrewsykim ?
And one concern: InstanceMetadataByProviderID checked instance existence, do we need to remove these logic?

**Review by andrewsykim — COMMENTED** [review]
> And one concern: InstanceMetadataByProviderID checked instance existence, do we need to remove these logic?

Let's remove this -- this was introduced in this release anyways

> I'd like to remove these broken test, because most of them are instance not exist, WDYT @andrewsykim ?

That's fine I think as long as we have the instance not exists tests in the cloud node lifecycle controller

**Comment by gongguan** [comment]
> Let's remove this -- this was introduced in this release anyways

I prefer do it later while we can ensure updateNodeAddress not check instance existence works well.

**Comment by andrewsykim** [comment]
> And one concern: InstanceMetadataByProviderID checked instance existence, do we need to remove these logic?

Looking at the code again I don't see where `InstanceMetadataByProviderID` checks for existence, can you show me where this is happening?

**Comment by gongguan** [comment]
InstanceExistsByProviderID's logic also done in InstanceMetadataByProviderID, many are duplicated with NodeAddressesByProviderID, such as get instance. For example
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go#L166

**Comment by andrewsykim** [comment]
Oh I see, but that's specific to the provider implementation. We should defer to the providers to update that and the original PR should not have implemented it for all the providers.

**Comment by andrewsykim** [comment]
So yes, let's worry about that in a follow-up PR

**Comment by gongguan** [comment]
Yeah, it's better that each cloud provider update their own InstanceMetadataByProviderID and switch on instancesV2.

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

Thanks for iterating on this @gongguan!

**Comment by andrewsykim** [comment]
/milestone v1.19

(should have had the milestone from the start)

**Comment by gongguan** [comment]
pod pending timeout
/test pull-kubernetes-integration

**Comment by wojtek-t** [comment]
Nice - I'm really glad to see it happened (nice finding after 2.5 weeks of vacation). I didn't expect this happening in 1.19 timeframe.

**Comment by gongguan** [comment]
It doesn't change current cloud-node-controller except for last commit(which removed instance existence check). I'm glad to see that nothing abnormal happened during these weeks.
